add validator FundingStreams to allow multiple commission payout destinations

avahowell opened the pull request:
Closes #256

hdevalence commented on pd/src/db/validators.sql:
```diff
@@ -1,6 +1,5 @@
     voting_power bigint NOT NULL,
-    commission_address varchar NOT NULL,
-    commission_rate bigint NOT NULL
+    funding_streams bytea NOT NULL
```
Rather than storing the funding streams opaquely, could we keep a first-class table in the database, so we can query it later if we want to do other stuff with it?  I'm not sure what the right way to structure the table would be, since I guess we want a many-to-one relationship between funding streams and validators (cc @aubrey-dyer)

aubrey-dyer replied:
simplest way might just be adding a funding streams table with a key or indexed column that points back to the associated validator

hdevalence commented on pd/src/staking.rs:
```diff
@@ -3,6 +3,17 @@ use serde::{Deserialize, Serialize};
+    /// The destinatination address for the funding stream..
+    pub address: Address,
+
+    /// The portion (in terms of basis points) of the validator's total commission that goes to
+    /// this funding stream.
+    pub rate_bps: u16,
```
If we specify each funding stream as a percentage of the validator's total commission, we have an extra invariant that the sum of the `rate_bps` must be 10,000.

Could we instead specify the rate for each funding stream in absolute terms, and let the validator's total commission be specified implicitly as the sum of the commission for each funding stream?

hdevalence replied:
(Alternatively, if that's what this is already supposed to be doing -- the total rate disappeared from the `Validator` structure, so maybe it is? -- I guess the comment should be changed to say that it's a portion of the total rewards for that validator's delegation pool.

hdevalence commented:
Could this PR also update the spec sections that describe the commission as being a single rate to reflect having multiple funding streams?

avahowell commented:
Switched to a first-class table for `validator_fundingstreams`, updated the docstrings about how funding stream rates work, and updated the validator rewards part of the protocol spec to match.

hdevalence commented on spec/stake/validator-rewards.md:
```diff
@@ -1,9 +1,15 @@
-their delegators $r_{v,e} = (1 - c_{v,e})r_e$, or equivalently $r_e = r_{v,e} +
-c_{v,e}r_e$.
+Validators declare a set of funding streams that comprise the destination of
+all of their staking rewards. Each funding stream contains a rate $ri \in
+[0,1]$ and a destination address $ai$.  The validator's total commission rate
+is defined as $c_{v,e} = \sum ri$, the sum of the rate of each funding stream.
```
not sure i follow the notation, what's the `i`?

avahowell replied:
the `ith` funding stream

hdevalence reviewed (approved):
LGTM mod spec notation (nonblocking)
